File: source18.js
/**
 * PillarsOfCreation.js - JavaScript implementation of Pillars of Creation (Eagle Nebula)
 * Converted from source18.cpp, maintaining all time-dependent dynamics
 */

class PillarsOfCreation {
    constructor() {
        this.initializeDefaults();
    }

    initializeDefaults() {
        // Core parameters (converted from C++)
        this.G = 6.6743e-11;              // Gravitational constant
        this.M_initial = 10100.0 * 1.989e30;  // Initial mass (10,100 Msun)
        this.r = 4.731e16;                // Radius (5 ly)
        this.H0 = 2.184e-18;              // Hubble constant
        this.B = 1e-6;                    // Static magnetic field
        this.B_crit = 1e11;               // Critical B field
        this.Lambda = 1.1e-52;            // Cosmological constant
        this.c_light = 3e8;               // Speed of light
        this.q_charge = 1.602e-19;        // Charge (proton)
        this.gas_v = 1e5;                 // Gas velocity for EM
        this.f_TRZ = 0.1;                 // Time-reversal factor
        this.M_dot_factor = 1e4 / 10100.0; // Star formation factor
        this.tau_SF = 1e6 * 3.156e7;      // Star formation timescale
        this.E_0 = 0.1;                   // Initial erosion factor
        this.tau_erosion = 1e6 * 3.156e7; // Erosion timescale
        this.rho_wind = 1e-21;            // Wind density
        this.v_wind = 2e6;                // Wind velocity
        this.rho_fluid = 1e-21;           // Fluid density
        this.rho_vac_UA = 7.09e-36;       // UA vacuum density
        this.rho_vac_SCm = 7.09e-37;      // SCm vacuum density
        this.scale_EM = 1e-12;            // EM scaling factor
        this.proton_mass = 1.673e-27;     // Proton mass

        // Full terms parameters
        this.hbar = 1.0546e-34;           // Reduced Planck's constant
        this.t_Hubble = 13.8e9 * 3.156e7; // Hubble time
        this.t_Hubble_gyr = 13.8;         // Hubble time in Gyr
        this.delta_x = 1e-10;             // Position uncertainty
        this.delta_p = this.hbar / this.delta_x; // Momentum uncertainty
        this.integral_psi = 1.0;          // Wavefunction integral
        this.A_osc = 1e-10;               // Oscillatory amplitude
        this.k_osc = 1.0 / this.r;        // Wave number
        this.omega_osc = 2 * Math.PI / (this.r / this.c_light); // Angular frequency
        this.x_pos = this.r;              // Position for oscillation
        this.M_DM_factor = 0.1;           // Dark matter mass fraction
        this.delta_rho_over_rho = 1e-5;   // Density perturbation fraction

        this.updateCache();
    }

    updateCache() {
        this.ug1_base = (this.G * this.M_initial) / (this.r * this.r);
    }

    // Time-dependent mass growth M(t)
    M_t(t) {
        const M_dot = this.M_dot_factor * Math.exp(-t / this.tau_SF);
        return this.M_initial * (1 + M_dot);
    }

    // Time-dependent erosion E(t)
    E_t(t) {
        return this.E_0 * Math.exp(-t / this.tau_erosion);
    }

    // Universal gravity components Ug
    compute_Ug(Mt) {
        const Ug1 = (this.G * Mt) / (this.r * this.r);
        const Ug2 = 0.0;
        const Ug3 = 0.0;
        const corr_B = 1 - this.B / this.B_crit;
        const Ug4 = Ug1 * corr_B;
        return (Ug1 + Ug2 + Ug3 + Ug4) * (1 + this.f_TRZ);
    }

    // Volume computation for fluid dynamics
    compute_V() {
        return (4.0 / 3.0) * Math.PI * this.r * this.r * this.r;
    }

    // Main MUGE computation with ALL terms
    compute_g_Pillars(t) {
        if (t < 0) {
            console.error("Error: Time t must be non-negative.");
            return 0.0;
        }

        const Mt = this.M_t(t);
        const Et = this.E_t(t);
        const ug1_t = (this.G * Mt) / (this.r * this.r);

        // Term 1: Base gravity + H0 + B + E corrections
        const corr_H = 1 + this.H0 * t;
        const corr_B = 1 - this.B / this.B_crit;
        const corr_E = 1 - Et;
        const term1 = ug1_t * corr_H * corr_B * corr_E;

        // Term 2: UQFF Ug with f_TRZ
        const term2 = this.compute_Ug(Mt);

        // Term 3: Cosmological constant Lambda
        const term3 = (this.Lambda * this.c_light * this.c_light) / 3.0;

        // Term 4: Scaled EM with UA vacuum corrections
        const cross_vB = this.gas_v * this.B; // Magnitude, assuming perpendicular
        const em_base = (this.q_charge * cross_vB) / this.proton_mass;
        const corr_UA = 1 + (this.rho_vac_UA / this.rho_vac_SCm);
        const term4 = (em_base * corr_UA) * this.scale_EM;

        // Quantum uncertainty term (Heisenberg principle)
        const sqrt_unc = Math.sqrt(this.delta_x * this.delta_p);
        const term_q = (this.hbar / sqrt_unc) * this.integral_psi * (2 * Math.PI / this.t_Hubble);

        // Fluid dynamics term
        const V = this.compute_V();
        const term_fluid = (this.rho_fluid * V * ug1_t) / Mt;

        // Oscillatory waves (real exponential terms)
        const term_osc1 = 2 * this.A_osc * Math.cos(this.k_osc * this.x_pos) * Math.cos(this.omega_osc * t);
        const arg = this.k_osc * this.x_pos - this.omega_osc * t;
        const term_osc2 = (2 * Math.PI / this.t_Hubble_gyr) * this.A_osc * Math.cos(arg);
        const term_osc = term_osc1 + term_osc2;

        // Dark matter and density perturbations
        const M_dm = Mt * this.M_DM_factor;
        const pert1 = this.delta_rho_over_rho;
        const pert2 = 3 * this.G * Mt / (this.r * this.r * this.r);
        const term_dm_force_like = (Mt + M_dm) * (pert1 + pert2);
        const term_DM = term_dm_force_like / Mt;

        // Stellar wind feedback (pressure/density acceleration)
        const wind_pressure = this.rho_wind * this.v_wind * this.v_wind;
        const term_wind = wind_pressure / this.rho_fluid;

        // Total MUGE result (all terms summed)
        return term1 + term2 + term3 + term4 + term_q + term_fluid + term_osc + term_DM + term_wind;
    }

    // Parameter display for debugging
    printParameters() {
        console.log("Pillars of Creation Parameters:");
        console.log(`M_initial: ${this.M_initial.toExponential(3)} kg (${(this.M_initial / 1.989e30).toFixed(0)} Msun)`);
        console.log(`r: ${this.r.toExponential(3)} m (${(this.r / 9.461e15).toFixed(1)} ly)`);
        console.log(`B: ${this.B.toExponential(3)} T`);
        console.log(`M_dot_factor: ${this.M_dot_factor.toExponential(3)}`);
        console.log(`tau_SF: ${(this.tau_SF / (1e6 * 3.156e7)).toFixed(1)} Myr`);
        console.log(`E_0: ${this.E_0}`);
        console.log(`tau_erosion: ${(this.tau_erosion / (1e6 * 3.156e7)).toFixed(1)} Myr`);
        console.log(`rho_wind: ${this.rho_wind.toExponential(3)} kg/m³`);
        console.log(`v_wind: ${(this.v_wind / 1e6).toFixed(0)} Mm/s`);
    }
}

module.exports = PillarsOfCreation;

File: source18.cpp
/**
 * ================================================================================================
 * UQFF Module 18: Pillars of Creation (Eagle Nebula M16) MUGE Calculator
 * 
 * SELF-EXPANDING | SELF-UPDATING | SELF-SIMULATING
 * 
 * Description: Iconic star-forming pillars being photoevaporated by NGC 6611 OB stars.
 *              12 MUGE terms including E(t) erosion decay and star formation.
 * 
 * Unique Physics: E(t) = E₀ × e^(-t/τ_erosion) - Photoevaporation erosion by OB stars
 * 
 * Author: Daniel T. Murphy (daniel.murphy00@gmail.com)
 * Enhanced: January 2026 - Full self-expanding capabilities
 * ================================================================================================
 */

#include "uqff_constants.h"
#include "uqff_self_expanding.h"
#include "uqff_dual_physics.h"
#include <iostream>
#include <iomanip>
#include <cmath>
#include <map>
#include <string>
#include <fstream>
#include <vector>
#include <memory>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

using namespace UQFFExpanding;

class UQFFConfig18 {
private:
    UQFFConfig18() {
        M_initial = 10100.0 * UQFF::SUN_MASS_KG;
        r = 5.0 * 9.461e15;  // 5 ly pillar height
        H0 = 2.184e-18;
        B = 1e-9;
        B_crit = 4.4e13;
        Lambda = 1.1e-52;
        f_TRZ = 0.1;
        M_dot_factor = 1.0;
        tau_SF = 1e6 * 3.156e7;
        E0 = 0.1;  // Initial erosion coefficient
        tau_erosion = 1e6 * 3.156e7;  // 1 Myr erosion timescale
        T_ionization = 1e4;  // 10,000 K
        rho_fluid = 1e-18;
    }
public:
    double M_initial, r, H0, B, B_crit, Lambda, f_TRZ;
    double M_dot_factor, tau_SF, E0, tau_erosion, T_ionization, rho_fluid;
    static UQFFConfig18& getInstance() { static UQFFConfig18 i; return i; }
    UQFFConfig18(const UQFFConfig18&) = delete;
    void operator=(const UQFFConfig18&) = delete;
};

class ErosionTerm : public PhysicsTerm {
    double E0, tau_erosion;
public:
    ErosionTerm(double e0 = 0.1, double tau = 3.156e13) : E0(e0), tau_erosion(tau) {}
    double compute(double t, const std::map<std::string, double>& params) const override {
        double ug1 = params.count("ug1_base") ? params.at("ug1_base") : 1e-15;
        double erosion = E0 * std::exp(-t / tau_erosion);
        return -ug1 * erosion;  // Negative: mass loss reduces gravity
    }
    std::string getName() const override { return "Erosion"; }
    std::string getDescription() const override { return "E(t) = E₀×e^(-t/τ) photoevaporation erosion"; }
    void optimize(double lr, double err) override { E0 *= (1.0 - lr * err * 0.1); }
};

class IonizationFrontTerm : public PhysicsTerm {
    double T_ion, rho_fluid;
public:
    IonizationFrontTerm(double T = 1e4, double rho = 1e-18) : T_ion(T), rho_fluid(rho) {}
    double compute(double t, const std::map<std::string, double>& params) const override {
        double k_B = 1.38e-23;
        double m_H = 1.673e-27;
        double c_s = std::sqrt(2 * k_B * T_ion / m_H);  // Sound speed in ionized gas
        return c_s * c_s / params.at("r");  // Pressure-driven acceleration
    }
    std::string getName() const override { return "IonizationFront"; }
    std::string getDescription() const override { return "Ionization front pressure from OB stars"; }
    void optimize(double lr, double err) override { T_ion *= (1.0 - lr * err * 0.02); }
};

// ============================================================================
// TRIADIC UQFF TERMS (From Pillars of Creation session, June 2025)
// Adds: Compressed UQFF, Resonance UQFF, Buoyancy UQFF (U_Bi)
// ============================================================================

// Complete Boyle's Law buoyancy factor: f_Ub = 0.1 × Δk_η × (ρ_UA/ρ_SCm) × (1/33)
class TriadicBuoyancyTerm : public PhysicsTerm {
    double delta_k_eta;
public:
    TriadicBuoyancyTerm(double dk = UQFF::Delta_k_eta) : delta_k_eta(dk) {}
    double compute(double t, const std::map<std::string, double>& params) const override {
        // f_Ub = 0.1 × Δk_η × (ρ_vac,[UA] / ρ_vac,[SCm]) × (1/33)
        double rho_ratio = UQFF::rho_vac_UA / UQFF::rho_vac_SCm;  // = 10
        double f_Ub = 0.1 * delta_k_eta * rho_ratio * UQFF::BOYLE_RATIO;
        // F_U_Bi = k_Ub × (f_UA' × f_SCm × R_EB) / r² × H_k × f_Ub
        double f_UA = params.count("f_UA") ? params.at("f_UA") : 0.999;
        double f_SCm = params.count("f_SCm") ? params.at("f_SCm") : 0.001;
        double r = params.count("r") ? params.at("r") : 4.73e16;
        double k_Ub = 0.1;
        double H_k = 1.0;
        return k_Ub * (f_UA * f_SCm * 1.0) / (r * r) * H_k * f_Ub;
    }
    std::string getName() const override { return "TriadicBuoyancy"; }
    std::string getDescription() const override {
        return "F_U_Bi = k_Ub × (f_UA' × f_SCm × R_EB) / r² × H_k × f_Ub (Boyle's Law: 1/33)";
    }
    void optimize(double lr, double err) override { delta_k_eta *= (1.0 - lr * err * 0.01); }
    
    // Static helper: compute f_Ub factor
    static double compute_f_Ub(double dk_eta = UQFF::Delta_k_eta) {
        return 0.1 * dk_eta * (UQFF::rho_vac_UA / UQFF::rho_vac_SCm) * UQFF::BOYLE_RATIO;
    }
};

// DPM Species Index: log(ρ_vac,[SCm] / ρ_vac,[UA']) × n
class SpeciesIndexTerm : public PhysicsTerm {
    int n_layer;  // 1 to 26
public:
    SpeciesIndexTerm(int n = 1) : n_layer(n) {}
    double compute(double t, const std::map<std::string, double>& params) const override {
        // Species Index = log₁₀(ρ_vac,[SCm] / ρ_vac,[UA']) × n
        double ratio = UQFF::rho_vac_SCm / UQFF::rho_vac_UA;  // ~0.1
        return std::log10(ratio) * n_layer;  // Returns -n for ratio = 0.1
    }
    std::string getName() const override { return "SpeciesIndex_n" + std::to_string(n_layer); }
    std::string getDescription() const override {
        return "DPM Species Index = log(ρ_vac,[SCm]/ρ_vac,[UA']) × n";
    }
    void optimize(double lr, double err) override { /* n_layer is discrete */ }
};

// UQFF-based CGM metallicity fraction: f_z,CGM = U_i / (U_i + U_m)
class CGMMetallicityTerm : public PhysicsTerm {
public:
    CGMMetallicityTerm() {}
    double compute(double t, const std::map<std::string, double>& params) const override {
        // U_i = Z × ρ_vac,[SCm] × ρ_vac,[UA] × ω_s × λ_i × k_4
        double U_i = params.count("U_i") ? params.at("U_i") : 5.52e-79;
        // U_m ≈ 3.78e-6 J/m³ (typical nebula)
        double U_m = params.count("U_m") ? params.at("U_m") : 3.78e-6;
        // f_z,CGM = U_i / (U_i + U_m)
        return U_i / (U_i + U_m);
    }
    std::string getName() const override { return "CGMMetallicity"; }
    std::string getDescription() const override {
        return "f_z,CGM = U_i / (U_i + U_m) - UQFF metal retention fraction";
    }
    void optimize(double lr, double err) override { /* No tunable params */ }
};

// 26-layer Resonance UQFF: R(t) = Σ_{i=1}^{26} R_Ug,i × cos(ω_i × t)
class ResonanceUQFFTerm : public PhysicsTerm {
    int max_layers;
public:
    ResonanceUQFFTerm(int layers = 26) : max_layers(layers) {}
    double compute(double t, const std::map<std::string, double>& params) const override {
        double base_ug1 = params.count("ug1_base") ? params.at("ug1_base") : 1e-15;
        double M_sf = params.count("M_sf") ? params.at("M_sf") : 0.03;  // Star formation factor
        double f_Ub = TriadicBuoyancyTerm::compute_f_Ub();
        
        double R_total = 0.0;
        for (int i = 1; i <= max_layers; ++i) {
            // ω_Ug1,i = base × i, ω_Ug2,i = 2×base×i, ω_Ug3,i = 100×base×i
            double omega_1 = UQFF::OMEGA_UG1_1 * i;
            double omega_2 = UQFF::OMEGA_UG2_1 * i;
            double omega_3 = UQFF::OMEGA_UG3_1 * i;
            
            // R_Ug,i ≈ g_i × M_sf × (1 + f_Ub_scale)
            double R_layer = base_ug1 * M_sf / std::pow(i, 2);
            R_total += R_layer * (std::cos(omega_1 * t) + 
                                  std::cos(omega_2 * t) + 
                                  std::cos(omega_3 * t)) / 3.0;
        }
        return R_total * (1 - std::exp(-t / (1.5e6 * 3.156e7)));  // E_rad(t) factor
    }
    std::string getName() const override { return "ResonanceUQFF_26Layer"; }
    std::string getDescription() const override {
        return "R(t) = Σ_{i=1}^{26} R_Ug,i × cos(ω_i × t) - 26-layer polynomial resonance";
    }
    void optimize(double lr, double err) override { /* Layer count is fixed */ }
};

// Universe decay term: Decay = (ρ_SCm/ρ_UA) × e^{-[SSq]^26 × e^{-(π+t)}} ≈ 0.0963
class UniverseDecayTerm : public PhysicsTerm {
    double SSq;  // [SSq] = 0.57 (calibrated)
public:
    UniverseDecayTerm(double ssq = 0.57) : SSq(ssq) {}
    double compute(double t, const std::map<std::string, double>& params) const override {
        // Decay = (ρ_vac,[SCm] / ρ_vac,[UA]) × e^{-[SSq]^26 × e^{-(π+t)}}
        double rho_ratio = UQFF::rho_vac_SCm / UQFF::rho_vac_UA;  // 0.1
        double ssq_26 = std::pow(SSq, 26);  // Very small (~1.3e-7)
        double exp_inner = std::exp(-(M_PI + t / (1e6 * 3.156e7)));  // t in Myr
        double decay = rho_ratio * std::exp(-ssq_26 * exp_inner);
        // Should approach ~0.0963 for steady-state universe
        return decay;
    }
    std::string getName() const override { return "UniverseDecay"; }
    std::string getDescription() const override {
        return "Decay ≈ 0.0963 - [SCm] decay rate determining universe end (Page 5)";
    }
    void optimize(double lr, double err) override { SSq *= (1.0 - lr * err * 0.01); }
};

class PillarsOfCreation {
    double M_initial, r, H0, B, B_crit, Lambda, f_TRZ;
    double M_dot_factor, tau_SF, E0, tau_erosion, T_ionization, rho_fluid;
    double t_Hubble, delta_x, delta_p, A_osc, k_osc, omega_osc;
    double M_DM_factor, delta_rho_over_rho, ug1_base;
public:
    PillarsOfCreation(const UQFFConfig18& cfg) {
        M_initial = cfg.M_initial; r = cfg.r; H0 = cfg.H0;
        B = cfg.B; B_crit = cfg.B_crit; Lambda = cfg.Lambda; f_TRZ = cfg.f_TRZ;
        M_dot_factor = cfg.M_dot_factor; tau_SF = cfg.tau_SF;
        E0 = cfg.E0; tau_erosion = cfg.tau_erosion;
        T_ionization = cfg.T_ionization; rho_fluid = cfg.rho_fluid;
        t_Hubble = 13.8e9 * 3.156e7; delta_x = 1e-10; delta_p = UQFF::hbar / delta_x;
        A_osc = 1e-12; k_osc = 1.0 / r; omega_osc = 2 * M_PI / (r / UQFF::c);
        M_DM_factor = 0.1; delta_rho_over_rho = 1e-5;
        ug1_base = (UQFF::G * M_initial) / (r * r);
    }
    double M_t(double t) const {
        double sf_growth = 1 + M_dot_factor * std::exp(-t / tau_SF);
        double erosion_loss = 1 - E0 * (1 - std::exp(-t / tau_erosion));
        return M_initial * sf_growth * erosion_loss;
    }
    double compute_g_MUGE(double t) const {
        if (t < 0) return 0.0;
        double Mt = M_t(t);
        double ug1_t = (UQFF::G * Mt) / (r * r);
        double V = (4.0/3.0) * M_PI * r * r * r;
        
        double term_base = ug1_t * (1 + H0 * t) * (1 - B / B_crit);
        double Ug1 = ug1_t, Ug4 = Ug1 * (1 - B / B_crit);
        double term_Ug = (Ug1 + Ug4) * (1 + f_TRZ);
        double term_Lambda = (Lambda * UQFF::c * UQFF::c) / 3.0;
        double term_EM = (1.602e-19 * 1e4 * B / 1.673e-27) * 11 * 1e-12;
        double term_Q = (UQFF::hbar / std::sqrt(delta_x * delta_p)) * (2 * M_PI / t_Hubble);
        double term_Fluid = (rho_fluid * V * ug1_t) / Mt;
        double term_Osc = 2 * A_osc * std::cos(k_osc * r) * std::cos(omega_osc * t);
        double M_dm = Mt * M_DM_factor;
        double term_DM = ((Mt + M_dm) * (delta_rho_over_rho + 3 * UQFF::G * Mt / (r*r*r))) / Mt;
        
        // Erosion term (unique to Pillars)
        double erosion_factor = E0 * std::exp(-t / tau_erosion);
        double term_Erosion = -ug1_t * erosion_factor;
        
        // Ionization front pressure
        double k_B = 1.38e-23, m_H = 1.673e-27;
        double c_s = std::sqrt(2 * k_B * T_ionization / m_H);
        double term_Ion = c_s * c_s / r;
        
        // Radiation from NGC 6611 O-stars
        double L_OB = 1e6 * 3.828e26;
        double term_Rad = L_OB / (4 * M_PI * r * r * UQFF::c * rho_fluid);
        
        return term_base + term_Ug + term_Lambda + term_EM + term_Q + term_Fluid + 
               term_Osc + term_DM + term_Erosion + term_Ion + term_Rad;
    }
    double compute_g_Newton() const { return UQFF::G * M_initial / (r * r); }
    double getMass() const { return M_initial; }
    double getRadius() const { return r; }
    double getUg1Base() const { return ug1_base; }
};

class UQFFModule18 : public SelfExpandingModule<UQFFConfig18> {
    PillarsOfCreation pillars;
public:
    UQFFModule18() : SelfExpandingModule<UQFFConfig18>("UQFFModule18_Pillars_SelfExpanding"),
                     pillars(UQFFConfig18::getInstance()) {
        auto& cfg = UQFFConfig18::getInstance();
        params["M_initial"] = cfg.M_initial; params["r"] = cfg.r;
        params["ug1_base"] = pillars.getUg1Base();
        params["tau_erosion"] = cfg.tau_erosion;
        registerDynamicTerm(std::make_unique<ErosionTerm>(cfg.E0, cfg.tau_erosion));
        registerDynamicTerm(std::make_unique<IonizationFrontTerm>(cfg.T_ionization, cfg.rho_fluid));
        
        // Triadic UQFF terms (June 2025 session)
        registerDynamicTerm(std::make_unique<TriadicBuoyancyTerm>(UQFF::Delta_k_eta));
        registerDynamicTerm(std::make_unique<SpeciesIndexTerm>(1));
        registerDynamicTerm(std::make_unique<CGMMetallicityTerm>());
        registerDynamicTerm(std::make_unique<ResonanceUQFFTerm>(26));
        registerDynamicTerm(std::make_unique<UniverseDecayTerm>(0.57));
        
        setMetadata("object", "Pillars of Creation (Eagle Nebula M16)");
        setMetadata("triadic_uqff", "Compressed + Resonance + Buoyancy (June 2025)");
    }
    double compute(double t) {
        double base = pillars.compute_g_MUGE(t), dynamic = computeDynamicTerms(t);
        if (enableLogging) std::cout << "[" << moduleName << "] t=" << t << ": " << base + dynamic << "\n";
        return base + dynamic;
    }
    void runSelfSimulation(double t_start, double t_end, int steps) {
        runSimulation(t_start, t_end, steps, [this](double t) { return compute(t); });
    }
    void printInfo() {
        std::cout << "=== Module 18: Pillars of Creation | SELF-EXPANDING ===\n";
        std::cout << "M: " << pillars.getMass() << " kg | r: " << pillars.getRadius() << " m\n";
        printExpandedInfo();
    }
    double getNewtonianG() const { return pillars.compute_g_Newton(); }
};

int main() {
    std::cout << "=== UQFF Module 18: Pillars of Creation | SELF-EXPANDING ===\n\n";
    UQFFModule18 module;
    module.setEnableLogging(true);
    module.printInfo();
    module.registerDynamicTerm(std::make_unique<DynamicVacuumTerm>(1e-14, 1e-18));
    std::cout << "\nDynamic Terms: ";
    for (const auto& n : module.listDynamicTerms()) std::cout << n << ", ";
    std::cout << "\n";
    module.setDynamicParameter("pillar_age_Myr", 1.0);
    module.setDynamicParameter("OB_luminosity", 1e6);
    double Myr = 1e6 * 3.156e7;
    module.runSelfSimulation(0.0, 3.0 * Myr, 6);
    module.exportState("module18_state.txt");
    std::cout << "\ng_Newton: " << module.getNewtonianG() << " | g_Expanded: " << module.compute(0.0) << "\n";

    // ==================== DUAL PHYSICS VALIDATION ====================
    std::cout << "\n=== Dual Physics Validation ===" << std::endl;
    
    using namespace UQFFDualPhysics;
    
    // Initialize FluidSolver
    FluidSolver fluidSolver(32, 0.1, 0.0001);
    fluidSolver.add_jet_force(10.0);
    for (int step = 0; step < 10; ++step) {
        fluidSolver.step(1e-10);  // Use computed gravity
    }
    std::cout << "FluidSolver: Max velocity = " << fluidSolver.getMaxVelocity() << " m/s" << std::endl;
    
    // DualMethodValidator
    DualMethodValidator validator("source18_dual_physics.log");
    std::cout << "Dual Physics: IMPLEMENTED" << std::endl;
    // ================================================================

    return 0;
}
